File: day3/main.cpp
#include <iostream>
#include <string>
#include <vector>
#include <fstream>
#include <deque>
#include "utils_aoc.h"


using namespace std;

bool in_grid(int i, int j, const vector<string> &lines) {
    return i >= 0 && j >= 0 && i < lines.size() && j < lines[i].size();
}

bool is_symbol(int i, int j, const vector<string> &lines) {
    return lines[i][j] != '.' && !isdigit(lines[i][j]);
}

bool is_not_banned(int i, int j, int ban_i, int ban_j) {
    return i != ban_i || j != ban_j;
}

bool is_adjacent(int i, int j, const vector<string> &lines) {
    vector<vector<int>> dirs{
            {0,  1},
            {0,  -1},
            {1,  0},
            {-1, 0},
            {1,  1},
            {1,  -1},
            {-1, -1},
            {-1, 1}
    };
    if (!isdigit(lines[i][j])) {
        cout << " not digit! " << endl;
        return false;
    }

    for (const auto &dir: dirs) {
        int row = i + dir[0];
        int col = j + dir[1];

        if (in_grid(row, col, lines) && is_symbol(row, col, lines)) {
            return true;
        }
    }

    return false;
}

void part1(ifstream &input) {
    string line;

    vector<string> lines;

    while (getline(input, line)) {
        lines.push_back(line);
    }

    int sum = 0;
    int n = lines.size();

    for (int i = 0; i < lines.size(); i++) {
        for (int j = 0; j < lines[0].size(); j++) {
            if (isdigit(lines[i][j])) {
                int x = stoi(lines[i].substr(j));
                bool adjacent = false;
                while (j < lines[i].size() && isdigit(lines[i][j])) {
                    if (!adjacent && is_adjacent(i, j, lines)) {
                        adjacent = true;
                        sum += x;
                        j++;
                    } else {
                        j++;
                    }
                }
            }
        }
    }

    cout << sum << endl;
}

int get_number(int row, int col, const vector<string> &lines, vector<vector<bool>> &visited) {
    int j = col;
    while (j >= 0 && isdigit(lines[row][j])) {
        visited[row][j] = true;
        j--;
    }
    j++;
    int x = stoi(lines[row].substr(j));
    while(j<=lines[row].size() && isdigit(lines[row][j])) {
        visited[row][j]=true;
        j++;
    }

    return x;
}

long long gear_ratio(int i, int j, const vector<string> &lines) {
    vector<vector<int>> dirs{
            {0,  1},
            {0,  -1},
            {1,  0},
            {-1, 0},
            {1,  1},
            {1,  -1},
            {-1, -1},
            {-1, 1}
    };

    int counter = 0;
    long long mult = 1;

    vector<vector<bool>> visited(lines.size(), vector<bool>(lines[0].size(), false));
    for (const auto &dir: dirs) {
        int row = i + dir[0];
        int col = j + dir[1];

        if (in_grid(row, col, lines) && isdigit(lines[row][col]) && !visited[row][col]) {
//            cout<< "; "<<row<< " "<< col<<" "<<lines[row][col]<<endl;
            counter++;
            if (counter <= 2) {
                mult *= get_number(row, col, lines, visited);
            }
        }
    }

    if (counter == 2) return mult;

    return 0;
}

void part2(ifstream &input) {
    string line;
    vector<string> lines;
    while (getline(input, line)) {
        lines.push_back(line);
    }
    long long sum = 0;

//    cout<<gear_ratio(8,5,lines)<<endl;
//    vector<vector<bool>> visited(lines.size(), vector<bool>(lines[0].size(), false));
//    cout<<get_number(7,6,lines,visited)<<endl;

    for (int i = 0; i < lines.size(); i++) {
        for (int j = 0; j < lines[0].size(); j++) {
            if (lines[i][j] == '*') {
                sum += gear_ratio(i, j, lines);
            }
        }
    }

    cout << sum << endl;
}

int main() {
    ifstream input("/home/leartsura/Documents/projects/aoc23/input");
    if (input.fail()) {
        cout << "cant open input file" << endl;
        exit(0);
    }

//    part1(input);
    part2(input);

    input.close();
}

File: day3/Main.java
import java.io.BufferedReader;
import java.io.FileReader;
import java.io.IOException;
import java.text.NumberFormat;
import java.text.ParseException;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

public class Main {

    public static int stoi(String s, int beginIndex) throws ParseException {
        NumberFormat nf = NumberFormat.getInstance();
        return nf.parse(s.substring(beginIndex)).intValue();
    }


    private static boolean isAdjacent(int i, int j, ArrayList<String> grid) {
        int[][] directions = {
                {0, 1},
                {1, 0},
                {0, -1},
                {-1, 0},
                {1, 1},
                {-1, -1},
                {1, -1},
                {-1, 1}
        };

        for (int[] move : directions) {
            int row = i + move[0];
            int col = j + move[1];

            if (inGrid(row, col, grid) && isSymbol(row, col, grid)) {
                return true;
            }
        }

        return false;
    }

    private static boolean isSymbol(int row, int col, ArrayList<String> grid) {
        return grid.get(row).charAt(col) != '.' && !Character.isDigit(grid.get(row).charAt(col));
    }


    private static boolean inGrid(int i, int j, ArrayList<String> grid) {
        return i >= 0 && j >= 0 && i < grid.size() && j < grid.get(i).length();
    }

    public static void part1(BufferedReader br) throws Exception {
        String line;

        ArrayList<String> grid = new ArrayList<>();
        while ((line = br.readLine()) != null) {
            grid.add(line);
        }


        int sum = 0;
        for (int i = 0; i < grid.size(); i++) {
            for (int j = 0; j < grid.get(i).length(); j++) {
                if (Character.isDigit(grid.get(i).charAt(j))) {
                    int x = stoi(grid.get(i), j);
                    boolean adjacent = false;
                    while (j < grid.get(i).length() && Character.isDigit(grid.get(i).charAt(j))) {
                        if (!adjacent && isAdjacent(i, j, grid)) {
                            adjacent = true;
                        }
                        j++;
                    }
                    if (adjacent) {
                        sum += x;
                    }
                }
            }
        }

        System.out.println(sum);
    }


    public static void part2(BufferedReader br) throws Exception {
        String line;
        ArrayList<String> grid = new ArrayList<>();
        while ((line = br.readLine()) != null) {
            grid.add(line);
        }

        long sum = 0;
        for (int i = 0; i < grid.size(); i++) {
            for (int j = 0; j < grid.get(i).length(); j++) {
                if (grid.get(i).charAt(j) == '*') {
                    sum += gearRatio(i, j, grid);
                }
            }
        }

        System.out.println(sum);
    }

    private static long gearRatio(int i, int j, ArrayList<String> grid) throws ParseException {
        int[][] directions = {
                {0, 1},
                {1, 0},
                {0, -1},
                {-1, 0},
                {1, 1},
                {-1, -1},
                {1, -1},
                {-1, 1}
        };


        boolean[][] visited = new boolean[grid.size()][grid.get(0).length()];
        long mult=1;
        int counter=0;
        for (int[] move : directions) {
            int row = i + move[0];
            int col = j + move[1];

            if (inGrid(row, col, grid) && Character.isDigit(grid.get(row).charAt(col)) && !visited[row][col]) {
                counter++;
                if(counter<3) {
                    int num = getValue(row, col, grid, visited);
                    mult*=num;
                }
            }
        }
        if(counter!=2) {
            return 0;
        }

        return mult;
    }

    private static int getValue(int row, int col, ArrayList<String> grid, boolean[][] visited) throws ParseException {
        int j = col;
        while (j >= 0 && Character.isDigit(grid.get(row).charAt(j))) {
            visited[row][j] = true;
            j--;
        }
        j++;
        int x = stoi(grid.get(row), j);

        while (j < grid.get(row).length() && Character.isDigit(grid.get(row).charAt(j))) {
            visited[row][j] = true;
            j++;
        }
        return x;
    }

    public static void main(String[] args) throws Exception {
        BufferedReader br = new BufferedReader(new FileReader("jaoc23/src/input"));
//        part1(br);
        part2(br);

    }
}